src/txn/txn_rollback_to_stable.c

```diff
@@ -110,20 +45,200 @@ __txn_abort_newer_update(
             upd->durable_ts = upd->start_ts = WT_TS_NONE;
         }
     }
+
+    /* Reset the history store flag for the stable update. */
+    if (first_upd != NULL)
+        F_CLR(first_upd, WT_UPDATE_HS);
+
+    return (first_upd);
 }
 
 /*
  * __txn_abort_newer_insert --
- *     Apply the update abort check to each entry in an insert skip list
+ *     Apply the update abort check to each entry in an insert skip list and return the first
+ *     satisfy update.
  */
-static void
+static WT_UPDATE *
 __txn_abort_newer_insert(
   WT_SESSION_IMPL *session, WT_INSERT_HEAD *head, wt_timestamp_t rollback_timestamp)
 {
     WT_INSERT *ins;
 
     WT_SKIP_FOREACH (ins, head)
-        __txn_abort_newer_update(session, ins->upd, rollback_timestamp);
+        if (ins->upd != NULL &&
+          __txn_abort_newer_update(session, ins->upd, rollback_timestamp) != NULL)
+            return (ins->upd);
+
+    return (NULL);
+}
+
+/*
+ * __txn_abort_row_replace_with_hs_value --
+ *     Abort updates in the history store and replace the on-disk with an update that satisfy than
+ *     the given timestamp.
+ */
+static int
+__txn_abort_row_replace_with_hs_value(
+  WT_SESSION_IMPL *session, WT_PAGE *page, WT_ROW *rip, wt_timestamp_t rollback_timestamp)
+{
+    WT_CURSOR *hs_cursor;
+    WT_DECL_ITEM(hs_key);
+    WT_DECL_ITEM(hs_value);
+    WT_DECL_ITEM(key);
+    WT_DECL_RET;
+    WT_PAGE_MODIFY *mod;
+    WT_TIME_PAIR hs_start, hs_stop;
+    WT_UPDATE *upd, **upd_entry;
+    wt_timestamp_t durable_ts;
+    size_t size;
+    uint32_t hs_btree_id, session_flags;
+    uint8_t prepare_state, type;
+    int cmp;
+    bool valid_update_found;
+
+    hs_cursor = NULL;
+    upd = NULL;
+    hs_btree_id = S2BT(session)->id;
+    session_flags = 0;
+    valid_update_found = false;
+
+    /* Allocate buffers for the data store and history store key. */
+    WT_RET(__wt_scr_alloc(session, 0, &key));
+    WT_ERR(__wt_scr_alloc(session, 0, &hs_key));
+    WT_ERR(__wt_scr_alloc(session, 0, &hs_value));
+
+    WT_ERR(__wt_row_leaf_key(session, page, rip, key, false));
+
+    /* Open a history store table cursor. */
+    __wt_hs_cursor(session, &hs_cursor, &session_flags);
+    ret = __wt_hs_cursor_position(session, hs_cursor, hs_btree_id, key, WT_TS_MAX);
+    for (; ret == 0; ret = hs_cursor->prev(hs_cursor)) {
```

Comment:
Walking backward with `prev` is somewhat unusual. A comment about where we start and why we're going backward would be helpful.

Reply:
Added comment to explain logic.